src/main/java/org/folio/search/service/consortium/ConsortiumInstanceSearchService.java

```diff
@@ -67,34 +89,105 @@ public ConsortiumItem getConsortiumItem(String id, CqlSearchRequest<Instance> se
+    }
+
     return new ConsortiumHoldingCollection()
-      .holdings(consortiumHoldings)
-      .totalRecords(consortiumHoldings.size());
+      .holdings(searchRecords.stream().map(SearchResult::getRecords).flatMap(List::stream).toList())
+      .totalRecords(searchRecords.iterator().next().getTotalRecords());
   }
 
-  public ConsortiumItemCollection fetchConsortiumBatchItems(CqlSearchRequest<Instance> searchRequest,
-                                                            Set<String> ids) {
-    var result = searchService.search(searchRequest);
-    var consortiumItems = result.getRecords().stream()
-      .flatMap(instance ->
-        instance.getItems().stream()
-          .filter(item -> ids.contains(item.getId()))
-          .map(item -> ConsortiumItemMapper.toConsortiumItem(instance.getId(), item))
-      )
-      .toList();
+  public ConsortiumItemCollection fetchConsortiumBatchItems(String tenant, Set<UUID> itemIds) {
+    validateIdsCount(itemIds.size());
+
+    var ids = itemIds.stream().map(UUID::toString).collect(Collectors.toSet());
+    var targetField = INSTANCE_ITEM_FIELD_NAME + ".id";
+
+    var searchRecords = getConsortiumBatchResults(tenant, ids, targetField, this::mapToConsortiumItem);
+
+    if (searchRecords.isEmpty()) {
+      return new ConsortiumItemCollection();
+    }
 
     return new ConsortiumItemCollection()
-      .items(consortiumItems)
-      .totalRecords(consortiumItems.size());
+      .items(searchRecords.stream().map(SearchResult::getRecords).flatMap(List::stream).toList())
+      .totalRecords(searchRecords.iterator().next().getTotalRecords());
+  }
+
+  private <T> List<SearchResult<T>> getConsortiumBatchResults(String tenant, Set<String> ids, String targetField,
+                                                              BiFunction<Instance, Set<String>, List<T>> recordMapper) {
+    var request = CqlSearchRequest.of(Instance.class, tenant, "", 0, 0, true, false, true);
+    var termsQuery = termsQuery(targetField, ids);
+
+    if (ids.size() < DEFAULT_MAX_SEARCH_RESULT_WINDOW) {
+      var searchSourceBuilder = queryBuilder(termsQuery, ids.size());
+      var response = searchRepository.search(request, searchSourceBuilder);
+      var searchResult = documentConverter.convertToSearchResult(response, request.getResourceClass(),
+        (hits, item) -> recordMapper.apply(item, ids));
+      var records = searchResult.getRecords().stream()
+        .flatMap(List::stream)
+        .toList();
+      return List.of(SearchResult.of(records.size(), records));
+    }
+
+    var searchSourceBuilder = queryBuilder(termsQuery, properties.getSearchConsortiumRecordsPageSize())
+      .sort(fieldSort(targetField).order(ASC))
+      .searchAfter(new Object[]{""});
+    var response = searchRepository.search(request, searchSourceBuilder);
+    List<SearchResult<T>> searchRecords = new ArrayList<>();
+
+    while (response.getHits() != null && response.getHits().getHits().length > 0) {
```

Comment:
Can't we use `searchRepository.msearch` instead, by partitioning `ids` using `DEFAULT_MAX_SEARCH_RESULT_WINDOW`?

Reply:
technically it is possible, but I think (deep) paginating with ```search_after``` is a solution tailored for this kind of cases. When we partition ```ids``` and run multiple query each query will be separate one, but with ```search_after``` each query will have necessary information to retrieve the next page of results.

Reply:
actually, here we are not dealing with an issue of deep pagination where there may be case of skipping over ```from``` number of hits and thus the need to keep in memory those documents, we are just overcoming the constraint of ```index.max_result_window``` which is by default set as 10_000. So, maybe partitioning the ```ids``` and running multiple queries (possibly with parallelism) would be faster, but I do not have any result to compare and say something now. After the merge I can test this approach on environment with more data and assess the run times.